src/cose_int.h

```diff
@@ -27,21 +82,80 @@ typedef struct CounterSign1 COSE_CounterSign1;
 #define _countof(x) (sizeof(x) / sizeof(x[0]))
 #endif
 
-typedef struct _COSE_KEY {
+#ifdef __cplusplus
+class COSE_KEY {
+   public:
 	int m_refCount;
 	cn_cbor *m_cborKey;
 	int m_flags;
-	struct _COSE_KEY *m_nextKey;
+	COSE_KEY *m_nextKey;
 #ifdef USE_CBOR_CONTEXT
 	cn_cbor_context m_allocContext;
 #endif
 #ifdef COSE_C_USE_OPENSSL
 	EVP_PKEY *m_opensslKey;
 #endif
 #ifdef COSE_C_USE_MBEDTLS
-	mbedtls_ecp_keypair * m_mbedtls_keypair;
+	mbedtls_ecp_keypair *m_mbedtls_keypair;
+#endif
+
+   public:
+	static COSE_KEY *KeysRoot;
+
+   public:
+	COSE_KEY()
+	{
+		m_refCount = 1;
+		m_nextKey = KeysRoot;
+		KeysRoot = this;
+	}
+
+	~COSE_KEY();
+
+	int AddRef() { return m_refCount += 1; }
```

Comment:
why no using shared pointers? this has issues with multi threading, shared pointers solved this already

Reply:
I was not sure how I could use shared pointers.  The problem is that we are returning the pointer to C code and then having it call back.  I thought that the shared-pointer code was designed for only having the pointer in the class and not moving it back and forth.